toxcore/group_connection.c

```diff
@@ -85,18 +85,32 @@ void gcc_set_recv_message_id(GC_Connection *gconn, uint64_t id)
 }
 
 /** @brief Puts packet data in array_entry.
+ *
+ * Requires an empty array entry to be passed, and must not modify the passed
+ * array entry on error.
  *
  * Return true on success.
  */
-non_null(1, 2) nullable(3)
-static bool create_array_entry(const Mono_Time *mono_time, GC_Message_Array_Entry *array_entry, const uint8_t *data,
-                               uint16_t length, uint8_t packet_type, uint64_t message_id)
+non_null(1, 2, 3) nullable(4)
+static bool create_array_entry(const Logger *log, const Mono_Time *mono_time, GC_Message_Array_Entry *array_entry,
+        const uint8_t *data, uint16_t length, uint8_t packet_type, uint64_t message_id)
 {
+    if (!array_entry_is_empty(array_entry)) {
+        LOGGER_WARNING(log, "Failed to create array entry; entry is not empty.");
+        return false;
+    }
+
     if (length == 0) {
+        if (data != nullptr) {
+            LOGGER_ERROR(log, "Got non-null data with zero length (type %d)", packet_type); // should never happen
```

Comment:
wouldn't this be an assert instead?

Reply:
What are the conditions where an assert is preferable to an error?

Reply:
an assert is used when we should be able to assume it is always the case. Which is generally the case for internal stuff like this.

Reply:
I'd say assert is for things that are (somewhat easily) statically provable. This is a dependent truth, not trivial. I'd probably keep it as ERROR, but if you want to make it stronger, there's LOGGER_FATAL, which has the added advantage over assert of having a format string.

Reply:
It's not an unrecoverable error. I don't see any reason to kill the process.

Reply:
The idea is that it would only die if some bug in toxcore exists, not that it is not unrecoverable.

Reply:
Alright I changed it to LOGGER_FATAL.